src/SprykerSdk/Zed/ComposerConstrainer/Business/Finder/UsedForeignModuleFinder/UsedForeignModuleFinder.php

```diff
@@ -0,0 +1,199 @@
+    public function __construct(ComposerConstrainerConfig $config, ComposerJsonReaderInterface $composerJsonReader)
+    {
+        $this->config = $config;
+        $this->composerJsonReader = $composerJsonReader;
+    }
+
+    /**
+     * @return \Generated\Shared\Transfer\UsedModuleCollectionTransfer
+     */
+    public function find(): UsedModuleCollectionTransfer
+    {
+        $usedModuleCollectionTransfer = new UsedModuleCollectionTransfer();
+
+        if (!is_dir($this->config->getSourceDirectory())) {
+            return $usedModuleCollectionTransfer;
+        }
+
+        foreach ($this->createFinder() as $splFileInfo) {
+            $usedModuleCollectionTransfer = $this->addUsedForeignModules($usedModuleCollectionTransfer, $splFileInfo);
+        }
+
+        return $usedModuleCollectionTransfer;
+    }
+
+    /**
+     * @return \Symfony\Component\Finder\Finder|\Symfony\Component\Finder\SplFileInfo[]
+     */
+    protected function createFinder(): Finder
+    {
+        $finder = new Finder();
+        $finder
+            ->files()
+            ->in($this->config->getSourceDirectory())
+            ->filter($this->getFileFilter())
+            ->exclude($this->config->getExcludedNamespaces())
+            ->name('*.php');
+
+        return $finder;
+    }
+
+    /**
+     * @return \Closure
+     */
+    protected function getFileFilter(): Closure
+    {
+        return function (SplFileInfo $fileInfo) {
+            $filePath = $fileInfo->getPathname();
+            if (preg_match('/src\/(.*?)\/(.*?)\/(.*?)Config.php|src\/(.*?)\/(.*?)\/(.*?)DependencyProvider.php/', $filePath)) {
+                return false;
+            }
+
+            return true;
+        };
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\UsedModuleCollectionTransfer $usedModuleCollectionTransfer
+     * @param \Symfony\Component\Finder\SplFileInfo $splFileInfo
+     *
+     * @return \Generated\Shared\Transfer\UsedModuleCollectionTransfer
+     */
+    protected function addUsedForeignModules(UsedModuleCollectionTransfer $usedModuleCollectionTransfer, SplFileInfo $splFileInfo): UsedModuleCollectionTransfer
+    {
+        $fileContent = $splFileInfo->getContents();
+        foreach ($this->getUsedClassesInFile($fileContent) as $usedClassName) {
+            if (!$this->isExcludedClass($usedClassName) && $this->isVendorClass($usedClassName)) {
+                $usedModuleTransfer = $this->getUsedModuleByClassName($usedClassName);
+                $usedModuleCollectionTransfer->addUsedModule($usedModuleTransfer);
+            }
+        }
+
+        return $usedModuleCollectionTransfer;
+    }
+
+    /**
+     * @param string $className
+     *
+     * @return bool
+     */
+    protected function isExcludedClass(string $className): bool
+    {
+        $namespaces = array_merge(
+            $this->config->getCoreNamespaces(),
+            $this->config->getExcludedNameSpaces(),
+            $this->config->getProjectNamespaces()
+        );
```

Comment:
This check will be executed a thousand times, maybe better to cache $namespaces?
Or maybe $pattern

Reply:
Discussed. Decided leave it as is, due to the fact that this data is already in memory and we would like to avoid adding any kind of state to the UsedForeignModuleFinder.